mmdet3d/models/middle_encoders/sparse_encoder_sassd.py

```diff
@@ -0,0 +1,321 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+import torch
+from mmcv.runner import auto_fp16
+from torch import nn as nn
+
+from mmdet3d.models.losses import (weighted_sigmoid_focal_loss,
+                                   weighted_smoothl1)
+from mmdet3d.ops import (SparseBasicBlock, make_sparse_convmodule,
+                         pts_in_boxes3d)
+from mmdet3d.ops import spconv as spconv
+from mmdet3d.ops import three_interpolate_2d, three_nn_2d
+from ..builder import MIDDLE_ENCODERS
+
+
+def tensor2points(tensor, offset=(0., -40., -3.), voxel_size=(.05, .05, .1)):
+    indices = tensor.indices.float()
+    offset = torch.Tensor(offset).to(indices.device)
+    voxel_size = torch.Tensor(voxel_size).to(indices.device)
+    indices[:, 1:] = (
+        indices[:, [3, 2, 1]] * voxel_size + offset + .5 * voxel_size)
+    return tensor.features, indices
+
+
+def nearest_neighbor_interpolate(unknown, known, known_feats):
+    """
+    :param pts: (n, 4) tensor of the bxyz positions of the unknown features
+    :par
+    am ctr: (m, 4) tensor of the bxyz positions of the known features
+    :param ctr_feats: (m, C) tensor of features to be propigated
+    :return:
+        new_features: (n, C) tensor of the features of the unknown features
+    """
+    dist, idx = three_nn_2d(unknown, known)
+    dist_recip = 1.0 / (dist + 1e-8)
+    norm = torch.sum(dist_recip, dim=1, keepdim=True)
+    weight = dist_recip / norm
+    interpolated_feats = three_interpolate_2d(known_feats, idx, weight)
+
+    return interpolated_feats
+
+
+@MIDDLE_ENCODERS.register_module()
+class SparseEncoderSASSD(nn.Module):
```

Comment:
Cane `SparseEncoderSASSD` be inherited from `SparseEncoder`? It seems lots of code in the class duplicates with the original `SparseEncoder`?

Reply:
and it seems this class can be directly added in to `sparse_encoder.py` file